test/functional/kernel/warp-thread/tests/warp-thread-ReduceMask-impl.hpp

```diff
@@ -0,0 +1,185 @@
+
+  allocateForallTestData<RAJA::Index_type>(directlen*looplen,
+                                     erased_work_res,
+                                     &work_array,
+                                     &check_array,
+                                     &test_array);
+
+  RAJA::ReduceMax<REDUCE_POL, int> max_thread(0);
+  RAJA::ReduceSum<REDUCE_POL, RAJA::Index_type> trip_count(0);
+  RAJA::ReduceSum<REDUCE_POL, RAJA::Index_type> worksum(0);
+
+  call_kernel<EXEC_POLICY, USE_RESOURCE>(
+                            RAJA::make_tuple(RAJA::TypedRangeSegment<RAJA::Index_type>(0, directlen), RAJA::TypedRangeSegment<RAJA::Index_type>(0, looplen)),
+                            work_res,
+                            [=] RAJA_DEVICE (RAJA::Index_type i, RAJA::Index_type j) {
+                              trip_count += 1;
+                              worksum += i; // i should only be 0..directlen-1
+                              max_thread.max(threadIdx.x);
+                            });
+
+  ASSERT_EQ(max_thread.get(), 255);
+  ASSERT_EQ(trip_count.get(), looplen*directlen);
+  ASSERT_EQ(worksum.get(), looplen*directlen*(directlen-1)/2);
+
+  deallocateForallTestData<RAJA::Index_type>(erased_work_res,
+                                       work_array,
+                                       check_array,
+                                       test_array);
+}
+
+template <typename WORKING_RES, typename EXEC_POLICY, typename REDUCE_POL, bool USE_RESOURCE>
+void KernelWarpThreadTest(const DEVICE_DEPTH_2_REDUCESUM_WARPMASK_FORI&,
+                          const RAJA::Index_type directlen,
+                          const RAJA::Index_type looplen)
+{
+  WORKING_RES work_res{WORKING_RES::get_default()};
+  camp::resources::Resource erased_work_res{work_res};
+
+  RAJA::Index_type* work_array;
+  RAJA::Index_type* check_array;
+  RAJA::Index_type* test_array;
+
+  allocateForallTestData<RAJA::Index_type>(directlen*looplen,
+                                     erased_work_res,
+                                     &work_array,
+                                     &check_array,
+                                     &test_array);
+
+  RAJA::ReduceMax<REDUCE_POL, int> max_thread(0);
+  RAJA::ReduceSum<REDUCE_POL, RAJA::Index_type> trip_count(0);
+  RAJA::ReduceSum<REDUCE_POL, RAJA::Index_type> worksum(0);
+
+  call_kernel_param<EXEC_POLICY, USE_RESOURCE>(
+                            RAJA::make_tuple(RAJA::TypedRangeSegment<RAJA::Index_type>(0, directlen), RAJA::TypedRangeSegment<RAJA::Index_type>(0, looplen)),
+                            RAJA::make_tuple((RAJA::Index_type)0, (RAJA::Index_type)0),
+                            work_res,
+                            [=] RAJA_DEVICE (RAJA::Index_type i, RAJA::Index_type j, RAJA::Index_type x, RAJA::Index_type y) {
+                              trip_count += 1;
+                              worksum += y; // y should only be 0..3
+                              max_thread.max(threadIdx.x);
```

Comment:
Looking ahead towards SYCL (assuming it gets supported) I don't think `threadIdx.x` will be the right syntax. Can't we use one of the variables listed above to get the threadIdx? This would enable us  to insulate programing model specifics inside of RAJA.

Reply:
That's a good point, this particular test case exercises CUDA & HIP thread masking specifically, which are not going to applicable to other things like SYCL. I was going to name this set of tests something like "gpu-specific", but figured that "warp-thread" would be a bit more accurate.

Reply:
Interesting, I figured since we could run SYCL on the GPU there would be something analogous, like threadIdx.x on a threadblock is a workItem on a workgroup.